cdap-proto/src/main/java/co/cask/cdap/proto/ProtoConstraint.java

```diff
@@ -0,0 +1,264 @@
+
+    protected final long millisAfterTrigger;
+
+    public DelayConstraint(long millisAfterTrigger) {
+      super(Type.DELAY);
+      this.millisAfterTrigger = millisAfterTrigger;
+      validate();
+    }
+
+    public long getMillisAfterTrigger() {
+      return millisAfterTrigger;
+    }
+
+    @Override
+    public void validate() {
+      validateInRange(millisAfterTrigger, "millisAfterTrigger", 1L, null);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      return this == o || o != null
+        && getClass() == o.getClass()
+        && getMillisAfterTrigger() == ((DelayConstraint) o).getMillisAfterTrigger();
+    }
+
+    @Override
+    public int hashCode() {
+      return (int) (getMillisAfterTrigger() ^ (getMillisAfterTrigger() >>> 32));
+    }
+
+    @Override
+    public String toString() {
+      return String.format("Delay(%d ms)", getMillisAfterTrigger());
+    }
+  }
+
+  /**
+   * Represents a last run constraint in REST requests/responses.
+   */
+  public static class LastRunConstraint extends ProtoConstraint {
+
+    protected final long millisSinceLastRun;
+
+    public LastRunConstraint(long millisSinceLastRun) {
+      super(Type.LAST_RUN);
+      this.millisSinceLastRun = millisSinceLastRun;
+      validate();
+    }
+
+    public long getMillisSinceLastRun() {
+      return millisSinceLastRun;
+    }
+
+    @Override
+    public void validate() {
+      validateInRange(millisSinceLastRun, "millisSinceLastRun", 1L, null);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      return this == o || o != null
+        && getClass() == o.getClass()
+        && getMillisSinceLastRun() == ((LastRunConstraint) o).getMillisSinceLastRun();
+    }
+
+    @Override
+    public int hashCode() {
+      return (int) (getMillisSinceLastRun() ^ (getMillisSinceLastRun() >>> 32));
+    }
+
+    @Override
+    public String toString() {
+      return String.format("LastRun(%d ms)", getMillisSinceLastRun());
+    }
+  }
+
+  /**
+   * Represents a time range constraint in REST requests/responses.
+   */
+  public static class TimeRangeConstraint extends ProtoConstraint {
+
+    private static final Pattern SIMPLE_TIME_PATTERN = Pattern.compile("^[0-9][0-9]:[0-9][0-9]$");
+
+    protected final String timeZone;
+    protected final String startTime;
+    protected final String endTime;
+
+    public TimeRangeConstraint(String startTime,
+                               String endTime,
+                               TimeZone timeZone) {
+      super(Type.TIME_RANGE);
+      this.startTime = startTime;
+      this.endTime = endTime;
+      this.timeZone = timeZone.getID();
+      validate();
+    }
+
+    public String getStartTime() {
+      return startTime;
+    }
+
+    public String getEndTime() {
+      return endTime;
+    }
+
+    public String getTimeZone() {
+      return timeZone;
+    }
+
+    @Override
+    public void validate() {
+      validateSimpleTime("start", startTime);
```

Comment:
This validation only happens when the constraint is specified in the REST API, but not from programmatic API, right (because the subclass has its own `validate` method)?
That seems a bit inconsistent and undesirable.